runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/data/DataUtil.java

```diff
@@ -224,20 +230,29 @@ public boolean hasNext() {
       }
       while (true) {
         try {
-          if (currentInputStream == null) {
+          if (encodedCountingStream == null) {
             if (inputStreams.hasNext()) {
-              currentInputStream = buildInputStream(inputStreams.next(), serializer.getStreamChainers());
+              serializedCountingStream = new CountingInputStream(inputStreams.next());
+              encodedCountingStream = new CountingInputStream(buildInputStream(
+                  serializedCountingStream, serializer.getStreamChainers()));
             } else {
               cannotContinueDecoding = true;
               return false;
             }
           }
-          next = serializer.getCoder().decode(currentInputStream);
+        } catch (final IOException e) {
+          throw new RuntimeException(e);
```

Comment:
Is it okay to throw `RuntimeException`? How about `BlockFetchException`?

Reply:
That catch block covers failures from `buildInputStream`, which will quite certainly fail after many retrials. Should we use `BlockFetchException` in this case?